internal/compiler/pathutils.rs

```diff
@@ -0,0 +1,528 @@
+                };
+            }
+            PC::DirectoryComponent(v) => clean_components.push(PC::DirectoryComponent(v)),
+            PC::FileComponent(v) => clean_components.push(PC::FileComponent(v)),
+        }
+    }
+    if clean_components.is_empty() {
+        ".".to_string()
+    } else {
+        let mut result = String::new();
+        for c in clean_components {
+            match c {
+                PC::RootComponent(v) => {
+                    result = v.to_string();
+                }
+                PC::EmptyComponent | PC::SameDirectoryComponent(_) => {
+                    unreachable!("Never in the vector!")
+                }
+                PC::ParentDirectoryComponent(v) => {
+                    result += &format!("{v}{separator}");
+                }
+                PC::DirectoryComponent(v) => result += &format!("{v}{separator}"),
+                PC::FileComponent(v) => {
+                    result += v;
+                }
+            }
+        }
+        result
+    }
+}
+
+pub fn dirname_string(path: &str) -> String {
+    let mut iter = component_iter(path);
+    let separator = iter.separator();
+    let mut result = String::new();
+
+    while let Some(component) = iter.next() {
+        match component {
+            PathComponent::RootComponent(v) => result = v.to_string(),
+            PathComponent::EmptyComponent => result.push(separator),
+            PathComponent::SameDirectoryComponent(v)
+            | PathComponent::ParentDirectoryComponent(v)
+            | PathComponent::DirectoryComponent(v) => result += &format!("{v}{separator}"),
+            PathComponent::FileComponent(_) => { /* nothing to do */ }
+        };
+    }
+
+    if result.is_empty() {
+        String::from(".")
+    } else {
+        result
+    }
+}
+
+pub fn dirname(path: &Path) -> PathBuf {
+    let Some(path_str) = path.to_str() else {
+        // URLs can always convert to string in Rust
+        return path.to_owned();
+    };
+
+    PathBuf::from(dirname_string(path_str))
+}
+
+pub fn clean_path(path: &Path) -> PathBuf {
+    let Some(path_str) = path.to_str() else {
+        // URLs can always convert to string in Rust
+        return path.to_owned();
+    };
+
+    if is_url(path) {
+        // URL is cleaned up while parsing!
+        let url = url::Url::parse(path_str).unwrap();
+        return PathBuf::from(url.to_string());
+    }
+
+    PathBuf::from(clean_path_string(path_str))
+}
+
+pub fn join(base: &Path, path: &Path) -> Option<PathBuf> {
+    if is_absolute(path) {
+        return Some(path.to_owned());
+    }
+
+    let Some(base_str) = base.to_str() else {
+        // URLs can always convert to string in Rust
+        return Some(path.to_owned());
+    };
+    let Some(path_str) = path.to_str() else {
+        // URLs can always convert to string in Rust
+        return Some(path.to_owned());
+    };
+
+    if is_url(base) {
+        let base_url = url::Url::parse(base_str).unwrap();
```

Comment:
is_url already parse the URL. and parsing the URL is not cheap. 
So better do  `if let Ok(base_url) = url::Url::parse(base_str)`